ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/resources/arjunacore/XAResourceRecord.java

```diff
@@ -211,11 +214,14 @@ public int topLevelPrepare() {
                 case XAException.XA_RBOTHER:
                 case XAException.XA_RBPROTO:
                 case XAException.XA_RBTIMEOUT:
-                    // No txn -> assume rollback
+                    if (associationEnded) {
+                        // No txn -> assume rollback
 
-                    // This will avoid calling rollback when aborting
-                    _rolledBack = true;
-                    return TwoPhaseOutcome.PREPARE_NOTOK;
+                        // This will avoid calling rollback when aborting
+                        _rolledBack = true;
+
+                        return TwoPhaseOutcome.PREPARE_NOTOK;
```

Comment:
Shouldn't the `return` instruction be outside the `if`?

Reply:
It could do, I was letting it drop through to the next section though to be handled with the XAER_INVAL/XAER_PROTO/XAER_RMERR/XAER_RMFAIL - what do you think?

Reply:
I think you are right, I'll update the pull request